### PR TITLE
[WIP] Created command for update python dependencies

### PR DESCRIPTION
**Pull request by tyulpan**
Created command in Makefile for update python dependencies #51

### CODE REVIEW

File: CONTRIBUTORS.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@
 ----------------
 Alexandr Hladyshenko
 Andriy Yatsynyak
+Andrey Zhabsky
```
Comment (webknjaz) [inline CONTRIBUTORS.txt]: Plz fix sorting

File: Makefile [hunk 2]
```diff
@@ -18,6 +18,13 @@ test:
 	doc8 project_new/docs/
 	cd project_new/ && docker-compose up test  && docker-compose stop
 
+upgrade-requirements:
+	cp requirements-dev.txt upgrade-requirements.txt
```
Comment (webknjaz) [inline Makefile]: I believe you shouldn't need a temporary file for this

File: create_aio_app/utils/check-requirements.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import filecmp
+
+if filecmp.cmp("requirements-dev.txt", "upgrade-requirements.txt") is False:
```
Comment (webknjaz) [inline create_aio_app/utils/check-requirements.py]:
```suggestion
if not filecmp.cmp("requirements-dev.txt", "upgrade-requirements.txt"):
```

File: requirements-dev.txt [hunk 4]
```diff
@@ -8,7 +8,7 @@ flake8-debugger==3.1.0
 flake8-mutable==1.2.0
 flake8-tuple==0.2.14
 pyroma==2.4
-click==7.0
+click==8.0
```
Comment (webknjaz) [inline requirements-dev.txt]: This change is unrelated

File: create_aio_app/utils/check-requirements.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import subprocess
+
+uniq_tools = ""
+
+with open("requirements-dev.txt", "r") as f:
```
Comment (Arfey) [inline create_aio_app/utils/check-requirements.py]: In this case u can iterate direcory's files

```python
import os
from pathlib import Path      

path = Path('./requirements') 

for file in os.listdir(path):
    ... 
```

File: create_aio_app/utils/check-requirements.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    old_dependencies = f.read().split('\n')
+    new_dependencies = []
+    for line in old_dependencies:
+        if line != '-e .' and line != '':
+            tool = line[0:line.find("-")] if line.find(
+                "-") > 0 else line[0:line.find("=")]
+            if uniq_tools != tool:
+                uniq_tools = tool
+                fresh_version = subprocess.check_output(
+                    "pip freeze | grep " + tool,
```
Comment (Arfey) [inline create_aio_app/utils/check-requirements.py]: pip freeze don't return a new version of package, may be do u mean 

```
pip list -o   
```
?

File: create_aio_app/utils/check-requirements.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                    shell=True).decode("utf-8").split('\n')[:-1]
+                new_dependencies += fresh_version
+
+with open("requirements-dev.txt", "w") as f:
+    for tool in new_dependencies:
+        f.write("%s\n" % tool)
+    f.write("-e .\n")
+
+if old_dependencies is not new_dependencies:
+    print(
```
Comment (Arfey) [inline create_aio_app/utils/check-requirements.py]: use color 🌈  output with `click` 😀

File: create_aio_app/utils/check-requirements.py (hunk 5 above)
Comment (webknjaz) [reply create_aio_app/utils/check-requirements.py]: I believe `pathlib` has built-in methods for that.

File: create_aio_app/template/{{cookiecutter.project_name}}/Makefile [hunk 8]
```diff
@@ -28,6 +28,10 @@ clean:
 bash:
 	@docker exec -it $(PROJECT_NAME) bash
 
+## Runs script for upgrade dependencies:
+upgrade:
+	python3 -m $(PROJECT_NAME).utils.check-requirements.py
```
Comment (Arfey) [inline create_aio_app/template/{{cookiecutter.project_name}}/Makefile]: looks much better 👍

File: create_aio_app/template/{{cookiecutter.project_name}}/docs/source/pages/commands.rst [hunk 9]
```diff
@@ -14,7 +14,8 @@ Common
    ``make run (make)``, To start the project in develop mode
    ``make stop``, To stop work of docker containers
    ``make stop``, To clean up of docker containers
-   ``make bash``, Interactive work inside container (the command must be running after ``make run`` )
+   ``make bash``, Interactive work inside container (the command must be running after ``make run`` 
+   ``make upgrade``, To upgrade dependencies)
```
Comment (Arfey) [inline create_aio_app/template/{{cookiecutter.project_name}}/docs/source/pages/commands.rst]: Problem with text's format (move the bracket back please 😀)

File: create_aio_app/template/{{cookiecutter.project_name}}/Makefile (hunk 8 above)
Comment (Arfey) [reply create_aio_app/template/{{cookiecutter.project_name}}/Makefile]: btw, we talked about running this command out from the docker to upgrade packages that related with a current project.
U can do that some like that:

*version 1*

`docker-compose.yml`
```yaml
upgrade:
    <<: *app
    container_name: {{ cookiecutter.project_name }}_upgrande
    command: python3 -m  {{ cookiecutter.project_name }}.utils.check-requirements.py
```
`Makefile`
```
upgrade:
 	docker-compose --rm run upgrade
```

*version 2*

or use only `Makefile`

`Makefile`
```
upgrade:
 	docker-compose --rm run app python3 -m $(PROJECT_NAME).utils.check-requirements.py
```

Personally, i prefer *version 2*, because we change only one file. Important part here is `--rm` don't forget to add that, please.

[docker-compose run - docs](https://docs.docker.com/compose/reference/run/)

File: create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import subprocess
+import os
+import click
+from functools import partial
+from pathlib import Path
+
```
Comment (Arfey) [inline create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py]: Here u can read reasons why need to write main function for python's script.

https://stackoverflow.com/questions/45129728/what-is-the-best-practice-for-running-a-python-script?answertab=votes#tab-top

I can add from me:

 - reusing a code 
 - a better structure of code

Of course, this does not affect the final result, but I highly recommend using this propose with `main ()` function 😀

File: create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+if __name__ == '__main__':
+    path = Path('../../requirements/')
+
+    for file in os.listdir(path):
```
Comment (Arfey) [inline create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py]: U use `Path` from `pathlib`, so u can use built-in method for it 😀

*simple example*

```python
import pathlib

requirements_dir = pathlib.Path('./requirements')


def update_requirements() -> None:
    req_files = [req_file for req_file in requirements_dir.iterdir() if req_file.is_file()]

    with req_file.open() as f:
        # do your stuff
```

[link to docs](https://docs.python.org/3/library/pathlib.html#basic-use)

File: create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        with open("../../requirements/" + file, "r") as text:
+            old_dependencies = text.read().split('\n')
+            new_dependencies = []
+
+            for line in old_dependencies:
```
Comment (Arfey) [inline create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py]: For me code is complicated and some time i don't know what is going on (as example `old_line.find('jinja2')`). I think, it's because u wanna coverage all cases. 

What do I suggest?

*first step* - get a new version of packages
```python
import subprocess

...
fresh_version = subprocess\
    .check_output("pip list -o --format=freeze", shell=True)\
    .decode("utf-8")\
    .split("\n")

packages = dict(package.split('==') for package in fresh_version if package)
```
result
```
print(packages)

>>> {'Markdown': '2.6.10',
>>> 'Rx': '1.6.0',
>>> ...
```
*second step* - rewrite versions for the each file

```python
with req_file.open() as f:
    text = f.read()
    for name, version packages.items():
        text = re.sub(f'{name}==[\d .]*', f'{name}=={version}', text) 
    # write to the file
    f.seek(0)
    f.write(text)
    f.truncate()
```

what do u think?

**Comment by Arfey** [comment]
@tyulpan you can do that, little left to finish 💪

File: create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py (hunk 12 above)
Comment (tyulpan) [reply create_aio_app/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/utils/check-requirements.py]: Yeah, the problem has changed, but the methods remain the same. Instead of rewriting the code after the first review, I tried to change it by overloading the solution. Classic. :smiley:
Your propose looks more elegant.:fire: I like it. 
Thanks for this review. I learned a lot of new things. :thumbsup:

File: create_aio_app/template/{{cookiecutter.project_name}}/docs/source/pages/commands.rst (hunk 9 above)
Comment (tyulpan) [reply create_aio_app/template/{{cookiecutter.project_name}}/docs/source/pages/commands.rst]: :man_facepalming:

**Comment by Arfey** [comment]
@tyulpan thx for your contribute 👍

**Comment by Arfey** [comment]
@webknjaz I think in future we can move to https://github.com/peritus/bumpversion

**Comment by webknjaz** [comment]
Why?

**Comment by Arfey** [comment]
Because `bumpversion` give similar result but out of the box, with test and with 12 contributors that support it :)

**Comment by webknjaz** [comment]
Similar to what exactly?

**Comment by Arfey** [comment]
sorry, I was wrong
